### PR TITLE
OBPIH-5087 Add validation on 0 and negative numbers in putaways and bin replenishment

### PR DESCRIPTION
**Pull request by alannadolny**

### CODE REVIEW

File: src/js/components/put-away/PutAwaySecondPage.jsx [hunk 1]
```diff
@@ -131,21 +131,36 @@ class PutAwaySecondPage extends Component {
...
+        let disabledMessage;
+
+        if (props.value < 1) {
+          disabledMessage = this.props.translate(
+            'react.putAway.negativeQuantity.label',
+            'Quantity cannot be lower than 1',
+          );
+        }
+
+        if (props.original && props.value > props.original.quantityAvailable) {
```
Comment (kchelstowski) [inline src/js/components/put-away/PutAwaySecondPage.jsx:143]: can be shorten to:
```
if (props.value > props.original?.quantityAvailable)
```

File: src/css/main.scss [hunk 2]
```diff
@@ -445,6 +445,14 @@ body {
...
   color: $red-beta;
 }
+div.has-error {
+  input {
+    border: 2px solid $red-beta;
+    -webkit-box-shadow: inset 0 1px 1px rgba(0, 0, 0, .075);
+    box-shadow: inset 0 1px 1px rgba(0, 0, 0, .075);
+    border-radius: 4px;
+  }
+}
```
Comment (kchelstowski) [inline src/css/main.scss:455]: why did you add those styles? Don't we have them already?

File: src/js/components/put-away/PutAwaySecondPage.jsx [hunk 3]
```diff
@@ -131,21 +131,36 @@ class PutAwaySecondPage extends Component {
...
+              html={disabledMessage}
+              disabled={!disabledMessage}
               theme="transparent"
               arrow="true"
               delay="150"
               duration="250"
               hideDelay="50"
             >
-              <div className={props.value > props.original.quantityAvailable ? 'has-error' : ''}>
+              <div className={props.value > props.original.quantityAvailable ||
```
Comment (kchelstowski) [inline src/js/components/put-away/PutAwaySecondPage.jsx:161]: I'd split this logic to a variables:
```
const isQuantityHigherThanOriginal = props?.value > props.original?.quantityAvailable
const isQuantityLowerThan1 = props?.value < 1
```
and use it above and also here inside `<div>`
```
<div className={(isQuantityHigherThanOriginal || isQuantityLowerThan1) ? 'has-error' : ''}>
```

File: src/js/components/put-away/PutAwaySecondPage.jsx [hunk 4]
```diff
@@ -164,18 +179,15 @@ class PutAwaySecondPage extends Component {
...
+            html={disabledMessage}
+            disabled={!disabledMessage}
             theme="transparent"
             arrow="true"
             delay="150"
             duration="250"
             hideDelay="50"
           >
-            <div className={props.original && props.value > props.original.quantityAvailable ? 'has-error' : ''}>
+            <div className={(props.original && props.value > props.original.quantityAvailable) || props.value < 1 ? 'has-error' : ''}>
```
Comment (kchelstowski) [inline src/js/components/put-away/PutAwaySecondPage.jsx:190]: apply the same logic from above here

File: src/js/components/put-away/PutAwaySecondPage.jsx [hunk 5]
```diff
@@ -625,6 +638,8 @@ class PutAwaySecondPage extends Component {
         }
         <div className="submit-buttons">
           <button
+            disabled={_.some(this.state.putAway.putawayItems, putawayItem =>
+                (putawayItem.quantity > putawayItem.quantityAvailable) || putawayItem.quantity < 1)}
```
Comment (kchelstowski) [inline src/js/components/put-away/PutAwaySecondPage.jsx:642]: I can see a similar logic here, but I assume it is not inside the `Cell`, so how about creating a function and pass there two arguments - the function that could be used for this case and above cases?

File: src/js/components/put-away/SplitLineModal.jsx [hunk 6]
```diff
@@ -132,10 +151,18 @@ class SplitLineModal extends Component {
...
    */
-  isValid() {
+
+  isNegativeQuantity() {
+    return _.some(this.state.splitItems, items => _.toInteger(items.quantity) <= 0);
+  }
+
+  isQuantityHigherThanOriginalPutaway() {
     const qtySum = this.calculatePutAwayQty();
+    return qtySum > _.toInteger(this.props.putawayItem.quantity);
```
Comment (kchelstowski) [inline src/js/components/put-away/SplitLineModal.jsx:161]: just want to follow - do you have to use `_.toInteger`, because the quantity is in a string format?

File: src/js/components/put-away/SplitLineModal.jsx [hunk 7]
```diff
@@ -132,10 +151,18 @@ class SplitLineModal extends Component {
...
+  }
 
-    return qtySum <= _.toInteger(this.props.putawayItem.quantity);
+  isValid() {
+    return !this.isNegativeQuantity() && !this.isQuantityHigherThanOriginalPutaway();
```
Comment (kchelstowski) [inline src/js/components/put-away/SplitLineModal.jsx:165]: do you use it anywhere? I see that in line `228` you delete `disabled={this.isValid()}` where it was used and I can't see any usage of this anymore?

**Review by kchelstowski — CHANGES_REQUESTED** [review]

File: src/css/main.scss (hunk 2 above)
Comment (alannadolny) [reply src/css/main.scss:455]: We had them, but in the putaway modal (split line button) styles should be applied on .has-error > input (this styles didn't work before). I am adding this tag to styles below, but border-radius has to be added separately.

File: src/js/components/put-away/SplitLineModal.jsx (hunk 6 above)
Comment (alannadolny) [reply src/js/components/put-away/SplitLineModal.jsx:161]: Yes, the quantity is in a string format

File: src/js/components/put-away/SplitLineModal.jsx (hunk 7 above)
Comment (alannadolny) [reply src/js/components/put-away/SplitLineModal.jsx:165]: I used it in line 328 `disabled={!this.isValid() || !this.isBinSelected()}`

**Review by kchelstowski — APPROVED** [review]

File: grails-app/i18n/messages.properties [hunk 8]
```diff
@@ -3322,7 +3322,9 @@ react.putAway.splitLine.label=Split line
 react.putAway.totalQty.label=Total QTY
 react.putAway.putAwayQty.label=Putaway QTY
 react.putAway.sumOfAll.label=Sum of all split items quantities cannot be higher than original putaway item quantity
+react.putAway.negativeSumOfAll.label=Items quantity cannot be lower than 1
```
Comment (jmiranda) [inline grails-app/i18n/messages.properties:3325]: We don't have to do this in this PR, but we should change all language around higher vs lower to greater than vs less than.

File: src/js/components/put-away/SplitLineModal.jsx (hunk 6 above)
Comment (kchelstowski) [reply src/js/components/put-away/SplitLineModal.jsx:161]: hard to say in this specific place why it is a string, but I know that in some places we have quantity as string. My blind guess is that initially the quantity is type of number (fetched from DB), but then whenever you edit the value, it becomes the string, that's why I believe in many places we have some kind of validation, if the value given is not a `NaN`. I think that @alannadolny will explain this specific case, as it is just my blind guess. Am I right @alannadolny ?

File: src/js/components/put-away/PutAwaySecondPage.jsx [hunk 9]
```diff
@@ -24,6 +24,10 @@ import 'react-table/react-table.css';
 
 const SelectTreeTable = (customTreeTableHOC(ReactTable));
 
+function isInvalid(quantity, quantityAvailable) {
```
[the remark this answers is not in the document]
Comment (kchelstowski) [reply src/js/components/put-away/PutAwaySecondPage.jsx:27]: yeah I agree, something didn't feel right to me about the naming also, but I didn't want to be nitpicky, but we should definitely follow the same convention everywhere

[the remark this answers is not in the document]
Comment (jmiranda) [reply src/js/components/put-away/PutAwaySecondPage.jsx:27]: You only have to be nitpicky until everyone is on the same page. Then it becomes convention / second nature.

File: src/js/components/put-away/SplitLineModal.jsx (hunk 6 above)
Comment (alannadolny) [reply src/js/components/put-away/SplitLineModal.jsx:161]: Yeah @kchelstowski you're right. The fetched quantity is type of number, after editing the value, it becomes the string:
Screenshot from 2023-01-04 23-53-20
Someone set types of this quantity in propTypes and set it to string and number before
Screenshot from 2023-01-05 00-00-04

**Review by drodzewicz — APPROVED** [review]

File: src/js/components/put-away/SplitLineModal.jsx [hunk 10]
```diff
@@ -64,6 +64,25 @@ class SplitLineModal extends Component {
     }
   }
 
+  getErrorMessage() {
+    if (this.isQuantityHigherThanOriginalPutaway()) {
```
Comment (drodzewicz) [inline src/js/components/put-away/SplitLineModal.jsx:68]: should probably be `isQuantityGreaterThanOriginalPutaway`

**Review by drodzewicz — CHANGES_REQUESTED** [review]

**Review by kchelstowski — APPROVED** [review]

**Review by drodzewicz — APPROVED** [review]
